x-pack/plugins/security_solution/public/common/hooks/__mocks__/use_license.ts

```diff
@@ -0,0 +1,11 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { createLicenseServiceMock } from '../../../../common/license/mocks';
+
+export const licenseService = createLicenseServiceMock();
```

Comment:
LicenseService is a concrete class. It should not need to be mocked. Best practice is to init the real class, and inject any needed/mocked state as necessary into the instance.

Reply:
A simpler approach can be using a spy for useLicense and return the mocked value for the specific describe.

Reply:
What do you mean by concrete class?
In order to do this better, we would need to mock the observables that come out of kibana and that this class takes in during `.start()`. We can track that as tech. debt if needed - but I'm not trying to test `LicenseService` class here (if that was the case - then yes, totally agree 😄 ).

I'm just trying to mock the instance created by the `useLicense` hook. The issue  is that the hook instantiates and uses a `LicenseService` instance that is also exported. That exported instance is then used during the UI's `Plugin.start()` to give it the licensing observable. Because of this, its difficult to avoid the mocking of the hook. the implementation approach would have to be changed to using a React context provider - similar to how we have others [context providers defined](https://github.com/elastic/kibana/blob/master/x-pack/plugins/security_solution/public/common/mock/endpoint/app_root_provider.tsx#L46-L56) and then we could instantiate and use a custom LicenseService instance in bootstrapping the UI app.

Reply:
ok, so part of it _is_ how the frontend is set up here.

Since the `frontend/plugin.tsx` is starting the service with `licenseService.start(plugins.licensing.license$);` where `licenseService` is the imported instance. You don't have great control to supplant with your own instance, or passing in an observable you control (replacing `plugins.licensing.license$`).

It's still not good to mock value classes (sorry this is what I meant, not concrete), _specifically_ when they are not under test. But the singleton you don't control makes it hard.

Unless you can think of a sane way to replace what's in `plugins.licensing.license$`. That's really all the control you need to set the 'license state' for a given test.

Reply:
yeah, another thing I don't like about this mocking of the singleton is that it creates state between tests - I just got a failure here in CI that I think was caused by that.

I'm going to create a refactor issue to revisit this.